internal/pkg/manifest/svc.go

```diff
@@ -39,7 +40,85 @@ type Service struct {
 
 // ServiceImage represents the service's container image.
 type ServiceImage struct {
-	Build *string `yaml:"build"` // Path to the Dockerfile.
+	Build BuildArgsOrString `yaml:"build"` // Path to the Dockerfile.
+}
+
+// Dockerfile returns the path to the service's Dockerfile, defaulting to
+// the filename Dockerfile if not otherwise specified.
+func (s *ServiceImage) Dockerfile() string {
+	// Prefer to use the "dockerfile" string in BuildArgs. Otherwise,
+	// use "Context/Dockerfile", then "BuildString", then "Dockerfile"
+	if s.Build.BuildArgs.Dockerfile != nil {
+		return aws.StringValue(s.Build.BuildArgs.Dockerfile)
+	}
+
+	var dfPath string
+	if s.Build.BuildString != nil {
+		dfPath = aws.StringValue(s.Build.BuildString)
+	} else if s.Build.BuildArgs.Context != nil {
+		// This is a degraded experience, should only seldom happen when Dockerfile isn't explicit.
+		dfPath = aws.StringValue(s.Build.BuildArgs.Context)
+		dfPath = filepath.Join(dfPath, "Dockerfile")
+	} else {
+		dfPath = "./Dockerfile"
+	}
+
+	return dfPath
+}
+
+// Context returns the build context directory if it exists, otherwise an empty string.
+func (s *ServiceImage) Context() string {
+	return aws.StringValue(s.Build.BuildArgs.Context)
+}
+
+// Args returns the args section, if it exists, to override args in the dockerfile.
+// Otherwise it returns an empty map.
+func (s *ServiceImage) Args() map[string]string {
+	return s.Build.BuildArgs.Args
+}
+
+// BuildArgsOrString is a custom type which supports unmarshaling yaml which
+// can either be of type string or type DockerBuildArgs.
+type BuildArgsOrString struct {
+	BuildString *string
+	BuildArgs   DockerBuildArgs
```

Comment:
There's a tradeoff here: since we prefer to use BuildArgs.Dockerfile as our first source of truth, making `.BuildArgs` a pointer would require that we check whether `.BuildArgs` is nil before checking `.BuildArgs.Dockerfile`. I preferred to keep the "leaf" members as pointers so I didn't have to deal with the additional validation.

Reply:
Souynds good!

Reply:
I think though that making this into a pointer helps with the validation you suggested above, so I'm going to incorporate it and just add the validation logic into the Context() and Dockerfile() methods.